controller/lib/pid/pid.cpp

```diff
@@ -18,238 +17,90 @@ limitations under the License.
 
-  if (inAuto) {
-    if (*myOutput > outMax)
-      *myOutput = outMax;
-    else if (*myOutput < outMin)
-      *myOutput = outMin;
+  output_sum_ = stl::clamp(output_sum_, out_min_, out_max_);
 
-    if (outputSum > outMax)
-      outputSum = outMax;
-    else if (outputSum < outMin)
-      outputSum = outMin;
+  float res = output_sum_;
+  if (p_term_ == ProportionalTerm::ON_ERROR) {
+    res += kp * error;
   }
-}
-
-/* SetMode(...)****************************************************************
- * Allows the controller Mode to be set to manual (0) or Automatic (non-zero)
- * when the transition from manual to auto occurs, the controller is
- * automatically initialized
- ******************************************************************************/
-void PID::SetMode(int Mode) {
-  bool newAuto = (Mode == AUTOMATIC);
-  if (newAuto && !inAuto) { /*we just went from manual to auto*/
-    PID::Initialize();
+  if (d_term_ == DifferentialTerm::ON_MEASUREMENT) {
+    res -= kd * dInput / samplesTimeChangeSec;
+  } else {
+    res += kd * (error - last_error_) / samplesTimeChangeSec;
   }
-  inAuto = newAuto;
-}
 
-/* Initialize()****************************************************************
- *	does all the things that need to happen to ensure a bumpless transfer
- *  from manual to automatic mode.
- ******************************************************************************/
-void PID::Initialize() {
-  outputSum = *myOutput;
-  lastInput = *myInput;
-  lastError = *mySetpoint - *myInput;
-  if (outputSum > outMax)
-    outputSum = outMax;
-  else if (outputSum < outMin)
-    outputSum = outMin;
-  // next sample time in now
-  nextSampleTime = Hal.now();
-  // last call time defined as now - SampleTime to enable computation on first
-  // call (user should call Compute() immediately after SetMode(Auto))
-  lastUpdateTime = Hal.now() - SampleTime;
+  *output = stl::clamp(res, out_min_, out_max_);
+
+  // Remember some variables for next time
+  last_input_ = input;
+  last_error_ = error;
+  last_update_time_ = now;
+  // when should we expect to perform our next output calculation
+  next_sample_time_ = next_sample_time_ + sample_time_;
+  return true;
 }
 
-/* SetControllerDirection(...)*************************************************
- * The PID will either be connected to a DIRECT acting process (+Output leads
- * to +Input) or a REVERSE acting process(+Output leads to -Input.)  we need to
- * know which one, because otherwise we may increase the output when we should
- * be decreasing.  This is called from the constructor.
- ******************************************************************************/
-void PID::SetControllerDirection(int Direction) {
-  if (inAuto && Direction != controllerDirection) {
-    kp = (0 - kp);
-    ki = (0 - ki);
-    kd = (0 - kd);
-  }
-  controllerDirection = Direction;
+void PID::SetSampleTime(Duration sample_time) {
+  next_sample_time_ = next_sample_time_ - sample_time_ + sample_time;
```

Comment:
This looks quite confusing, I know it is correct, but it could look clearer.

Reply:
I removed this setter, now it can only be set in constructor.